### PR TITLE
Key typespecs

### PR DESCRIPTION
**Pull request by cjab**
Add typespecs to the rest of Diplomat.

### CODE REVIEW

File: config/test.exs [hunk 1]
```diff
@@ -5,3 +5,6 @@ config :goth,
 
 config :diplomat,
        token_module: Diplomat.TestToken
+
+config :ex_unit,
+       capture_log: true
```
Comment (cjab) [inline config/test.exs]: Bypass was pretty noisy during tests. This quiets it.

File: lib/diplomat/key.ex [hunk 2]
```diff
@@ -175,18 +154,37 @@ defmodule Diplomat.Key do
...
+  defp path_to_proto([], acc), do: acc
+  defp path_to_proto([[kind, id]|tail], acc) when is_nil(id) do
+    path_to_proto(tail, [PbPathElement.new(kind: kind, id_type: nil)|acc])
+  end
+  defp path_to_proto([[kind, id]|tail], acc) when is_integer(id) do
+    path_to_proto(tail, [PbPathElement.new(kind: kind, id_type: {:id, id})|acc])
+  end
+  defp path_to_proto([[kind, name]|tail], acc) do
+    path_to_proto(tail, [PbPathElement.new(kind: kind, id_type: {:name, name})|acc])
   end
```
Comment (cjab) [inline lib/diplomat/key.ex:167]: Sorry for the churn here 😟. I've grouped private functions at the bottom of the file. If you would prefer I didn't it's not a problem I can roll it back.

File: lib/diplomat/key_utils.ex [hunk 3]
```diff
@@ -2,23 +2,22 @@ defmodule KeyUtils do
...
-      _ ->
-        {:error, "Invalid data"}
-    end
+    key =
+      value
+      |> Base.url_decode64!(padding: false)
+      |> decode
+    {:ok, key}
+  rescue
+    _ -> {:error, "Invalid data"}
```
Comment (cjab) [inline lib/diplomat/key_utils.ex:20]: This just removes an unnecessary try.

File: config/test.exs (hunk 1 above)
Comment (peburrows) [reply config/test.exs]: Is this still necessary? I updated bypass yesterday (as part of `v0.8.0`), and it seems to have gotten rid of the noise.

File: lib/diplomat/key.ex [hunk 4]
```diff
@@ -143,29 +129,22 @@ defmodule Diplomat.Key do
...
+  def from_allocate_ids_proto(%AllocateIdsResponse{keys: keys}) do
     keys |> Enum.map(&from_proto(&1))
   end
 
+  @spec allocate_ids(String.t, pos_integer) :: [t] | Client.error
   def allocate_ids(type, count \\ 1) do
-    keys = Enum.map 1..count, fn(_i) ->
-      __MODULE__.new(type) |> __MODULE__.proto
+    keys = Enum.map 1..count, fn _ ->
+      type |> new() |> proto()
```
Comment (peburrows) [inline lib/diplomat/key.ex:140]: No idea why I was doing silly things like `__MODULE__.new` and `__MODULE__.proto`...

File: config/test.exs (hunk 1 above)
Comment (cjab) [reply config/test.exs]: Ooops! Yep, you're right it's no longer necessary. Gone!

**Comment by peburrows** [comment]
Thanks, @cjab! I've just pushed to hex as `0.8.2`.

**Pull request merged**